File: src/c#/Creational/FactoryMethod/FactoryMethod1/Employee.cs
using System;
using System.Text;

namespace FactoryMethod1
{
    public class Employee
    {
        public int no { get; set; }
        public String name { get; set; }
        public int year { get; set; }
        public string department { get; set; }
        public double salary { get; set; }

        public static readonly int BASE_SALARY = 500;

        public Employee(int no, String name, int year, string department)
        {
            this.no = no;
            this.name = name;
            this.year = year;
            this.department = department;
        }

        public virtual void Work()
        {
            Console.WriteLine("Employee is working!");
        }

        public virtual double CalculateSalary()
        {
            salary = year * BASE_SALARY;
            return salary;
        }

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder("\nEmployee Info");
            builder.Append("No: " + no);
            builder.Append("Name:" + name);
            builder.Append("Year: " + year);
            builder.Append("Department: " + department);
            builder.Append("Salary: " + CalculateSalary());
            return builder.ToString();
        }
    }
}

File: src/java/org/javaturk/dp/ch02/factory/problem/Employee.java
package org.javaturk.dp.ch02.factory.problem;

public class Employee {
	private static final int BASE_SALARY = 500;
	protected int no;
	protected String name;
	protected int year;
	protected String department;

	private String type;

	public Employee(int no, String name, int year, String department) {
		this.no = no;
		this.name = name;
		this.year = year;
		this.department = department;
	}

	public Employee(int no, String name, int year) {
		this.no = no;
		this.name = name;
		this.year = year;
	}

	public void work() {
		System.out.println("Employee is working!");
	}

	public void manage() {
		if (type.equals("Director"))
			System.out.println("Director is managing the whole company!");
		else if (type.equals("Manager"))
			System.out.println("Manager is managing his/her department!");
		else
			System.out.println("I am just a poor Employee and can only manage myself!");
	}

	public void makeAStrategicPlan() throws IamNotDirectorException {
		if (type.equals("Director"))
			System.out.println("Director is naking a strategic plan for the company!");
		else
			throw new IamNotDirectorException();
		
	}

	public void printInfo() {
		System.out.println("\nNo: " + no);
		System.out.println("Name:" + name);
		System.out.println("Year: " + year);
		System.out.println("Department: " + department);
	}

	@Override
	public String toString() {
		return "Employee [no=" + no + ", name=" + name + ", year=" + year + ", department=" + department + "]";
	}

	public double calculateSalary() {
		return year * BASE_SALARY;
	}
}
